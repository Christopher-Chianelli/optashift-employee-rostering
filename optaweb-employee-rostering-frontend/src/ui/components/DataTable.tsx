/*
 * Copyright 2019 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { PropsWithChildren, useState } from 'react';
import { Button, Pagination, Level, LevelItem } from '@patternfly/react-core';
import {
  SaveIcon, CloseIcon, EditIcon, TrashIcon, LongArrowAltDownIcon,
  LongArrowAltUpIcon, ArrowsAltVIcon,
} from '@patternfly/react-icons';
import { Sorter } from 'types';
import { useTranslation } from 'react-i18next';
import { setPropsInUrl, UrlProps } from 'util/BookmarkableUtils';
import { RouteComponentProps } from 'react-router';
import { PagenationData } from 'util/FunctionalComponentUtils';
import { TableComposable } from '@patternfly/react-table';
import FilterComponent from './FilterComponent';

export type DataTableUrlProps = UrlProps<'page'|'itemsPerPage'|'filter'|'sortBy'|'asc'>;

export const RowViewButtons = (props: { onEdit: () => void; onDelete: () => void }) => (
  <td role="cell">
    <span
      style={{
        display: 'grid',
        gridTemplateColumns: '1fr auto auto',
        gridColumnGap: '5px',
      }}
    >
      <span />
      <span>
        <Button
          variant="link"
          onClick={props.onEdit}
        >
          <EditIcon />
        </Button>
      </span>
      <span>
        <Button
          variant="danger"
          onClick={props.onDelete}
        >
          <TrashIcon />
        </Button>
      </span>
    </span>
  </td>
);

export const RowEditButtons = (props: { isValid: boolean; onClose: () => void; onSave: () => void }) => (
  <td role="cell">
    <span
      style={{
        display: 'grid',
        gridTemplateColumns: '1fr auto auto',
        gridColumnGap: '5px',
      }}
    >
      <span />
      <span>
        <Button
          variant="link"
          isDisabled={!props.isValid}
          onClick={() => {
            props.onSave();
            props.onClose();
          }}
        >
          <SaveIcon />
        </Button>
      </span>
      <span>
        <Button
          variant="link"
          onClick={props.onClose}
        >
          <CloseIcon />
        </Button>
      </span>
    </span>
  </td>
);

export const TableRow = (props: PropsWithChildren<{}>) => (<tr role="row">{props.children}</tr>);
export const TableCell = (props: PropsWithChildren<{}>) => (<td role="cell">{props.children}</td>);

export const PagenationControls = (props: PagenationData<any> & RouteComponentProps & {
  isCreatingNewRow: boolean;
  onCreateNewRow: () => void;
}) => {
  const { t } = useTranslation();
  return (
    <Level
      hasGutter
      style={{
        padding: '5px 5px 5px 5px',
        backgroundColor: 'var(--pf-global--BackgroundColor--200)',
      }}
    >
      <LevelItem>
        <FilterComponent
          filterText={props.filterText}
          onChange={(newFilterText) => {
            setPropsInUrl(props, { page: '1', filter: newFilterText });
          }}
        />
      </LevelItem>
      <LevelItem style={{ display: 'flex' }}>
        <Button
          isDisabled={props.isCreatingNewRow}
          onClick={props.onCreateNewRow}
        >
          {t('add')}
        </Button>
        <Pagination
          itemCount={props.numOfFilteredRows}
          perPage={props.itemsPerPage}
          page={props.page}
          onSetPage={(e, newPage) => setPropsInUrl(props, { page: `${newPage}` })}
          widgetId="pagination-options-menu-top"
          onPerPageSelect={(e, newPerPage) => setPropsInUrl(props, { itemsPerPage: `${newPerPage}` })}
        />
      </LevelItem>
    </Level>
  );
};

export function setSorterInUrl(props: RouteComponentProps, urlProps: { asc: string | null },
  oldIndex: number, newIndex: number) {
  if (oldIndex === newIndex) {
    setPropsInUrl(props, {
      asc: (urlProps.asc === 'true') ? 'false' : 'true',
    });
  } else {
    setPropsInUrl(props, {
      sortBy: `${newIndex}`,
      asc: 'true',
    });
  }
}
export interface TheTableProps<T> {
  title: string;
  columns: { name: string; sorter?: Sorter<T>}[];
  sortByIndex: number;
  rowWrapper: (item: T) => JSX.Element;
  newRowWrapper?: (removeRow: () => void) => JSX.Element;
  onSorterChange: (sortByColumnIndex: number) => void;
  // Only use this if you want to override add button behavior to NOT inline edit a row
  onAddButtonClick?: () => void;
}

/* eslint-disable no-nested-ternary */
// Patternfly Table in React is very inconvivent when the content is not text (ex: inline editing),
// so we use the HTML example to build it
export const TheTable = (props: PagenationData<any> & RouteComponentProps & TheTableProps<any>) => {
  const [isCreatingNewRow, setIsCreatingNewRow] = useState(false);

  return (
    <>
      <PagenationControls
        {...props}
        isCreatingNewRow={isCreatingNewRow}
        onCreateNewRow={() => {
          if (props.onAddButtonClick) {
            props.onAddButtonClick();
          } else {
            setIsCreatingNewRow(true);
          }
        }}
      />
      <TableComposable>
        <caption>{props.title}</caption>
        <thead>
          <tr role="row">
            {props.columns.map((header, index) => (
              <th
                key={header.name}
                role="columnheader"
                scope="col"
                className={(header.sorter !== undefined)
                  ? ((props.sortByIndex === index)
                    ? 'pf-c-table__sort pf-m-selected' : 'pf-c-table__sort'
                  )
                  : undefined
                }
                aria-sort={(header.sorter) ? (
                  (index === props.sortByIndex)
                    ? ((props.isReversed) ? 'descending' : 'ascending')
                    : 'none'
                )
                  : undefined}
              >
                <Button
                  variant="plain"
                  onClick={() => {
                    if (header.sorter !== undefined) {
                      props.onSorterChange(index);
                    }
                  }}
                >
                  <div className="pf-c-table__button-content">
                    <span className="pf-c-table__text">{header.name}</span>
                    {(header.sorter)
                      ? (
                        <span className="pf-c-table__sort-indicator">
                          {(index === props.sortByIndex)
                            ? ((props.isReversed) ? <LongArrowAltUpIcon /> : <LongArrowAltDownIcon />)
                            : (<ArrowsAltVIcon />)
                          }
                        </span>
                      ) : undefined
                    }
                  </div>
                </Button>
              </th>
            ))}
            <th role="columnheader" scope="col" />
          </tr>
        </thead>
        <tbody>
          {(isCreatingNewRow && props.newRowWrapper) ? props.newRowWrapper(() => {
            setIsCreatingNewRow(false);
          }) : undefined}
          {props.rowsInPage.map(item => props.rowWrapper(item)).toArray()}
        </tbody>
      </TableComposable>
    </>
  );
};
/* eslint-enable no-nested-ternary */
