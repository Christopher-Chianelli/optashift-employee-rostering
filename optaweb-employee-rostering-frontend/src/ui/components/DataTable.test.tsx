/*
 * Copyright 2019 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import * as React from 'react';
import { stringSorter } from 'util/CommonSorters';
import { getRouterProps } from 'util/BookmarkableTestUtils';
import { usePageableData } from 'util/FunctionalComponentUtils';
import { TheTable, TheTableProps } from './DataTable';

interface MockData {name: string; number: number}


const tableProps: TheTableProps<any> = {
  title: 'Data Table',
  columns: [{ name: 'Column 1' }, { name: 'Column 2' }],
  sortByIndex: 0,
  onSorterChange: jest.fn(),
  rowWrapper: jest.fn(),
};

const exampleData = [
  { name: 'Some Data', number: 1 },
  { name: 'More Data', number: 2 },
];

const routerProps = getRouterProps('/table', {});
const pageInfo = {
  page: '1',
  itemsPerPage: '5',
  filter: '',
  sortBy: '0',
  asc: 'true',
};

const useTableRows = (tableData: MockData[]) => usePageableData<MockData>(pageInfo,
  tableData, data => [data.name, `${data.number}`],
  stringSorter(data => data.name));

describe('DataTable component', () => {
  it('should render correctly with no rows', () => {
    const dataTable = shallow(<TheTable {...tableProps} {...routerProps} {...useTableRows([])} />);
    expect(toJson(dataTable)).toMatchSnapshot();
  });

  it('should render correctly with a few rows', () => {
    const dataTable = shallow(<TheTable {...tableProps} {...routerProps} {...useTableRows(exampleData)} />);
    expect(toJson(dataTable)).toMatchSnapshot();
  });
});
