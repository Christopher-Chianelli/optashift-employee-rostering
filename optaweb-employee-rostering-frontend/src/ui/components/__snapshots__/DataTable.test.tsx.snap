// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataTable component should render correctly with a few rows 1`] = `
<Fragment>
  <PagenationControls
    columns={
      Array [
        Object {
          "name": "Column 1",
        },
        Object {
          "name": "Column 2",
        },
      ]
    }
    filterText=""
    filteredRows={
      Immutable.List [
        Object {
          "name": "More Data",
          "number": 2,
        },
        Object {
          "name": "Some Data",
          "number": 1,
        },
      ]
    }
    history={
      Object {
        "action": "PUSH",
        "block": [MockFunction],
        "createHref": [MockFunction],
        "go": [MockFunction],
        "goBack": [MockFunction],
        "goForward": [MockFunction],
        "length": 1,
        "listen": [MockFunction],
        "location": Object {
          "hash": "",
          "pathname": "/table",
          "search": "",
          "state": undefined,
        },
        "push": [MockFunction],
        "replace": [MockFunction],
      }
    }
    isCreatingNewRow={false}
    isReversed={true}
    itemsPerPage={5}
    location={
      Object {
        "hash": "",
        "pathname": "/table",
        "search": "",
        "state": undefined,
      }
    }
    match={
      Object {
        "isExact": true,
        "params": Object {},
        "path": "/table",
        "url": "localhost:8080/table",
      }
    }
    numOfFilteredRows={2}
    onCreateNewRow={[Function]}
    onSorterChange={[MockFunction]}
    page={1}
    rowWrapper={
      [MockFunction] {
        "calls": Array [
          Array [
            Object {
              "name": "More Data",
              "number": 2,
            },
          ],
          Array [
            Object {
              "name": "Some Data",
              "number": 1,
            },
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      }
    }
    rowsInPage={
      Immutable.List [
        Object {
          "name": "More Data",
          "number": 2,
        },
        Object {
          "name": "Some Data",
          "number": 1,
        },
      ]
    }
    sortByIndex={0}
    title="Data Table"
  />
  <TableComposable>
    <caption>
      Data Table
    </caption>
    <thead>
      <tr
        role="row"
      >
        <th
          key="Column 1"
          role="columnheader"
          scope="col"
        >
          <Button
            onClick={[Function]}
            variant="plain"
          >
            <div
              className="pf-c-table__button-content"
            >
              <span
                className="pf-c-table__text"
              >
                Column 1
              </span>
            </div>
          </Button>
        </th>
        <th
          key="Column 2"
          role="columnheader"
          scope="col"
        >
          <Button
            onClick={[Function]}
            variant="plain"
          >
            <div
              className="pf-c-table__button-content"
            >
              <span
                className="pf-c-table__text"
              >
                Column 2
              </span>
            </div>
          </Button>
        </th>
        <th
          role="columnheader"
          scope="col"
        />
      </tr>
    </thead>
    <tbody />
  </TableComposable>
</Fragment>
`;

exports[`DataTable component should render correctly with no rows 1`] = `
<Fragment>
  <PagenationControls
    columns={
      Array [
        Object {
          "name": "Column 1",
        },
        Object {
          "name": "Column 2",
        },
      ]
    }
    filterText=""
    filteredRows={Immutable.List []}
    history={
      Object {
        "action": "PUSH",
        "block": [MockFunction],
        "createHref": [MockFunction],
        "go": [MockFunction],
        "goBack": [MockFunction],
        "goForward": [MockFunction],
        "length": 1,
        "listen": [MockFunction],
        "location": Object {
          "hash": "",
          "pathname": "/table",
          "search": "",
          "state": undefined,
        },
        "push": [MockFunction],
        "replace": [MockFunction],
      }
    }
    isCreatingNewRow={false}
    isReversed={true}
    itemsPerPage={5}
    location={
      Object {
        "hash": "",
        "pathname": "/table",
        "search": "",
        "state": undefined,
      }
    }
    match={
      Object {
        "isExact": true,
        "params": Object {},
        "path": "/table",
        "url": "localhost:8080/table",
      }
    }
    numOfFilteredRows={0}
    onCreateNewRow={[Function]}
    onSorterChange={[MockFunction]}
    page={1}
    rowWrapper={[MockFunction]}
    rowsInPage={Immutable.List []}
    sortByIndex={0}
    title="Data Table"
  />
  <TableComposable>
    <caption>
      Data Table
    </caption>
    <thead>
      <tr
        role="row"
      >
        <th
          key="Column 1"
          role="columnheader"
          scope="col"
        >
          <Button
            onClick={[Function]}
            variant="plain"
          >
            <div
              className="pf-c-table__button-content"
            >
              <span
                className="pf-c-table__text"
              >
                Column 1
              </span>
            </div>
          </Button>
        </th>
        <th
          key="Column 2"
          role="columnheader"
          scope="col"
        >
          <Button
            onClick={[Function]}
            variant="plain"
          >
            <div
              className="pf-c-table__button-content"
            >
              <span
                className="pf-c-table__text"
              >
                Column 2
              </span>
            </div>
          </Button>
        </th>
        <th
          role="columnheader"
          scope="col"
        />
      </tr>
    </thead>
    <tbody />
  </TableComposable>
</Fragment>
`;
